File: drivers/wlan/wlan/timer.cpp
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "timer.h"

#include <utility>

namespace wlan {

Timer::Timer(uint64_t id) : id_(id) {}

Timer::~Timer() {}

zx_status_t Timer::SetTimer(zx::time deadline) {
    deadline_ = deadline;
    return SetTimerImpl(deadline);
}

zx_status_t Timer::CancelTimer() {
    deadline_ = zx::time();
    return CancelTimerImpl();
}

SystemTimer::SystemTimer(uint64_t id, zx::timer timer) : Timer(id), timer_(std::move(timer)) {}

zx_status_t SystemTimer::SetTimerImpl(zx::time deadline) {
    return timer_.set(deadline, zx::duration());
}

zx_status_t SystemTimer::CancelTimerImpl() {
    return timer_.cancel();
}

zx_status_t TestTimer::SetTimerImpl(zx::time duration) {
    return ZX_OK;
}

zx_status_t TestTimer::CancelTimerImpl() {
    return ZX_OK;
}

}  // namespace wlan

File: garnet/bin/omaha_client/src/timer.rs
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use fuchsia_async as fasync;
use futures::future::BoxFuture;
use futures::prelude::*;
use omaha_client::state_machine::Timer;
use std::time::Duration;

pub struct FuchsiaTimer;

impl Timer for FuchsiaTimer {
    fn wait(&mut self, delay: Duration) -> BoxFuture<'static, ()> {
        fasync::Timer::new(fasync::Time::after(delay.into())).boxed()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::task::Poll;

    #[test]
    fn test_timer() {
        let mut exec = fasync::Executor::new().unwrap();
        let start_time = fasync::Time::now();

        let mut timer = FuchsiaTimer;
        let mut future = timer.wait(Duration::from_secs(1234));
        assert_eq!(Poll::Pending, exec.run_until_stalled(&mut future));

        let future_time = exec.wake_next_timer().unwrap();
        assert_eq!(1234, (future_time - start_time).into_seconds());

        assert_eq!(Poll::Ready(()), exec.run_until_stalled(&mut future));
    }
}
